auth/src/api/builder.rs

```diff
@@ -0,0 +1,122 @@
+    pub user_manager: UserManager,
+}
+
+pub struct ApiBuilder {
+    router: Router<RouterState>,
+    sqlite_pool: Option<Pool<Sqlite>>,
+}
+
+impl ApiBuilder {
+    pub fn new() -> Self {
+        let router = Router::new()
+            .route("/login", post(login))
+            .route("/logout", post(logout))
+            .route("/auth/session", post(convert_cookie))
+            .route("/auth/key", post(convert_key))
+            .route("/auth/refresh", post(refresh_token))
+            .route("/public-key", get(get_public_key))
+            .route("/user/:account_name", get(get_user).post(post_user))
+            .route_layer(from_extractor::<Metrics>())
+            .layer(
+                TraceLayer::new(|request| {
+                    request_span!(request, request.params.account_name = field::Empty)
+                })
+                .with_propagation()
+                .build(),
+            );
+
+        Self {
+            router,
+            sqlite_pool: None,
+        }
+    }
+
+    /// Connect and migrate an SQLite pool at the given URI.
+    pub async fn with_sqlite_pool(mut self, db_uri: &str) -> Self {
+        let sqlite_options = SqliteConnectOptions::from_str(db_uri)
+            .unwrap()
+            .create_if_missing(true)
+            // To see the sources for choosing these settings, see:
+            // https://github.com/shuttle-hq/shuttle/pull/623
+            .journal_mode(SqliteJournalMode::Wal)
+            .synchronous(SqliteSynchronous::Normal);
+
+        let pool = SqlitePool::connect_with(sqlite_options).await.unwrap();
+
+        MIGRATIONS.run(&pool).await.unwrap();
+
+        self.sqlite_pool = Some(pool);
+
+        self
+    }
+
+    /// Insert an admin user into the SQLite pool.
+    pub async fn init_db(mut self, args: InitArgs) -> Self {
+        let pool = self.sqlite_pool.expect("an sqlite pool is required");
+
+        let key = match args.key {
+            Some(ref key) => Key::from_str(key).unwrap(),
+            None => Key::new_random(),
+        };
+
+        query("INSERT INTO users (account_name, key, account_tier) VALUES (?1, ?2, 'admin')")
+            .bind(&args.name)
+            .bind(&key)
+            .execute(&pool)
+            .await
+            .expect("failed to insert admin user");
+
+        println!("`{}` created as admin user with key: {key}", args.name);
+
+        self.sqlite_pool = Some(pool);
+        self
+    }
+
+    pub fn into_router(self) -> Router {
+        let pool = self.sqlite_pool.expect("an sqlite pool is required");
+
+        let user_manager = UserManager { pool };
+        self.router.with_state(RouterState { user_manager })
+    }
+
+    pub async fn serve(self, address: SocketAddr) {
+        let router = self.into_router();
+
+        Server::bind(&address)
+            .serve(router.into_make_service())
+            .await
+            .unwrap_or_else(|_| panic!("Failed to bind to address: {}", address));
+    }
```

Comment:
If we keep these two separate then maybe the test can call router (from `into_router`) directly and therefore don't need to start up on a port? You can then also reach into the router state to set an admin user

Reply:
Good idea, that is a lot simpler, thanks!